dapr_agents/observability/wrappers/llm.py

```diff
@@ -0,0 +1,418 @@
+        display including tool calls and conversation context.
+
+        Args:
+            wrapped (callable): Original LLM generate method to be instrumented
+            instance (Any): LLM client instance containing model and configuration
+            args (tuple): Positional arguments - typically (messages,) for chat completions
+            kwargs (dict): Keyword arguments including model, tools, temperature, etc.
+
+        Returns:
+            Any: Result from wrapped method execution with comprehensive span attributes
+                 capturing LLM interaction details, token usage, and tool call information
+        """
+        # Check for instrumentation suppression
+        if context_api and context_api.get_value(
+            context_api._SUPPRESS_INSTRUMENTATION_KEY
+        ):
+            return wrapped(*args, **kwargs)
+
+        # Extract method parameters and build attributes
+        attributes = self._build_llm_attributes(wrapped, instance, args, kwargs)
+
+        # Handle async vs sync execution
+        if asyncio.iscoroutinefunction(wrapped):
+            return self._handle_async_execution(wrapped, args, kwargs, attributes)
+        else:
+            return self._handle_sync_execution(wrapped, args, kwargs, attributes)
+
+    def _build_llm_attributes(
+        self, wrapped: Any, instance: Any, args: Any, kwargs: Any
+    ) -> Dict[str, Any]:
+        """
+        Build comprehensive LLM span attributes with OpenInference formatting.
+
+        Constructs detailed span attributes including input messages, tool schemas,
+        model information, and serialized parameters following OpenInference
+        standards for proper Phoenix UI visualization and trace analysis.
+
+        Args:
+            wrapped (callable): Original LLM method for parameter binding
+            instance (Any): LLM client instance with model and configuration attributes
+            args (tuple): Positional arguments containing messages and parameters
+            kwargs (dict): Keyword arguments with model, tools, and LLM parameters
+
+        Returns:
+            Dict[str, Any]: Comprehensive span attributes including OpenInference-formatted
+                           input messages, tool schemas, model metadata, and serialized parameters
+        """
+        # Extract method parameters
+        messages = args[0] if args else kwargs.get("messages")
+        model = kwargs.get("model") or getattr(instance, "model", None)
+        tools = kwargs.get("tools") or []
+        agent_name = getattr(instance, "agent_name", None) or "ChatClient"
+
+        # Build base span attributes
+        attributes = {
+            OPENINFERENCE_SPAN_KIND: LLM,
+            INPUT_MIME_TYPE: "application/json",
+            OUTPUT_MIME_TYPE: "application/json",
+            "agent.name": agent_name,
+        }
+
+        if model:
+            attributes[LLM_MODEL_NAME] = model
```

Comment:
what if a user is running an app with say 3 agents within it and each using a diff model. How will this work if it is a single global const?

Reply:
That is a good question. I have not tested it with multiple agents and different modules.

Reply:
I tested this scenario too :) and same everything is captured the right way: 

<img width="1355" height="952" alt="image" src="https://github.com/user-attachments/assets/1ee08e65-6552-4c90-ac43-9817f9bfc5e6" />

<img width="1354" height="945" alt="image" src="https://github.com/user-attachments/assets/2576bbcf-e33c-4cdb-9c87-35a1c3455fce" />

<img width="1350" height="952" alt="image" src="https://github.com/user-attachments/assets/f38fde8c-e9b2-49aa-a11c-35d02b587393" />